File: CompilationTest/InheritanceTest/results/InheritanceTest.java
interface IPrinterBase {
  Integer someBaseFunc() throws Exception;
}

interface IPrinter extends IPrinterBase {
  void printIt() throws Exception;
}

class BasePrinter implements IPrinter {
    public Integer numValue = 42;

    public String getValue() throws Exception
    {
        return "Base";
    }
    
    public void printIt() throws Exception
    {
        System.out.println("BasePrinter: " + this.getValue());
    }
    
    public Integer someBaseFunc() throws Exception
    {
        return this.numValue;
    }
}

class ChildPrinter extends BasePrinter {
    public String getValue() throws Exception
    {
        return "Child";
    }
}

class TestClass {
    public IPrinter getPrinter(String name) throws Exception
    {
        IPrinter result = name.equals("child") ? new ChildPrinter() : new BasePrinter();
        return result;
    }
    
    public void testMethod() throws Exception
    {
        IPrinter basePrinter = this.getPrinter("base");
        IPrinter childPrinter = this.getPrinter("child");
        basePrinter.printIt();
        childPrinter.printIt();
        System.out.println(basePrinter.someBaseFunc() + " == " + childPrinter.someBaseFunc());
        
        BasePrinter baseP2 = new BasePrinter();
        ChildPrinter childP2 = new ChildPrinter();
        System.out.println(baseP2.numValue + " == " + childP2.numValue);
    }
}

class Program {
    public static void main(String[] args) throws Exception {
        try {
            new TestClass().testMethod();
        } catch (Exception err) {
            System.out.println("Exception: " + err.getMessage());
        }
    }
}

File: CompilationTest/InheritanceTest/results/InheritanceTest.cpp
#include <iostream>
#include <string>

class IPrinterBase {
  public:
    virtual int someBaseFunc() = 0;
};

class IPrinter : public IPrinterBase {
  public:
    virtual void printIt() = 0;
};

class BasePrinter : public IPrinter {
  public:
    int num_value = 42;

    virtual std::string getValue() {
        return std::string("Base");
    }
    
    void printIt() {
        std::cout << std::string("BasePrinter: ") + this->getValue() << std::endl;
    }
    
    int someBaseFunc() {
        return this->num_value;
    }

  private:
};

class ChildPrinter : public BasePrinter {
  public:
    std::string getValue() {
        return std::string("Child");
    }

  private:
};

class TestClass {
  public:
    one::sp<IPrinter> getPrinter(std::string name) {
        auto result = name == std::string("child") ? std::make_shared<ChildPrinter>() : std::make_shared<BasePrinter>();
        return result;
    }
    
    void testMethod() {
        auto base_printer = this->getPrinter(std::string("base"));
        auto child_printer = this->getPrinter(std::string("child"));
        base_printer->printIt();
        child_printer->printIt();
        std::cout << std::string(std::to_string(base_printer->someBaseFunc())) + " == " + std::to_string(child_printer->someBaseFunc()) << std::endl;
        
        auto base_p2 = std::make_shared<BasePrinter>();
        auto child_p2 = std::make_shared<ChildPrinter>();
        std::cout << std::string(std::to_string(base_p2->num_value)) + " == " + std::to_string(child_p2->num_value) << std::endl;
    }

  private:
};

int main()
{
    try {
        TestClass c;
        c.testMethod();
    } catch(std::exception& err) {
        std::cout << "Exception: " << err.what() << '\n';
    }
    return 0;
}